src/librustc_mir_build/lib.rs

```diff
@@ -0,0 +1,27 @@
+//! Construction of MIR from HIR.
+//!
+//! This crate also contains the match exhaustiveness and usefulness checking.
```

Comment:
Let's follow up with moving this out from here as well (into two crates, one for HAIR and one for match checking)? It feels unrelated to MIR building other than the HAIR structure as an in-between IR.

Reply:
Match "exhaustiveness" and "usefulness" calculations here are probably fairly bad description for what is being done in that code. They do not do any analysis that result in user-visible messages or anything of the sort. Rather that analysis is performed entirely as the part of – and for the purpose of – building MIR graph that only contains the relevant and necessary branches.

The analysis and the code building portions are heavily interleaved and I don’t believe splitting just for the sake of splitting is meaningful here. Especially now that `librustc_mir_build` is a relatively tiny crate already.

Reply:
> Match "exhaustiveness" and "usefulness" calculations referred to here are probably fairly bad description for what is being done in that code.

On the contrary, they feel like apt descriptions of the analysis that `hair::{_match, check_match}` and friends are doing.

> They do not do any analysis that result in user-visible messages or anything of the sort. 

Match checking certainly results in emitting plenty of its own error messages whether it be from exhaustiveness checking or other types of analysis in `check_match`.

> Rather that analysis is performed entirely as the part of – and for the purpose of – building MIR graph that only contains the relevant and necessary branches.

I don't think that's correct. MIR is in fact not sound by construction due to not verifying the exhaustiveness of `match`es which is done on HAIR, which I think is a language in its own right. One could say that rustc_typeck is only performed for the purpose of eventual MIR output, but that doesn't make HIR less of a language.
 
> The analysis and the code building portions are heavily interleaved and I don’t believe splitting just for the sake of splitting is meaningful here. Especially now that `librustc_mir_build` is a relatively tiny crate already.

I grepped quickly and searching for `crate::build` seems to generate zero hits in `src/librustc_mir/hair/` so this doesn't seem to be accurate?